lib/database_functions.rb

```diff
@@ -0,0 +1,56 @@
+module DatabaseFunctions
+  API_RATE_LIMIT = %(
+    CREATE OR REPLACE FUNCTION api_rate_limit(user_id int8)
+      RETURNS int4
+      AS $$
+    DECLARE
+      min_changes_per_hour int4 := #{Settings.min_changes_per_hour};
+      initial_changes_per_hour int4 := #{Settings.initial_changes_per_hour};
+      max_changes_per_hour int4 := #{Settings.max_changes_per_hour};
+      days_to_max_changes int4 := #{Settings.days_to_max_changes};
+      importer_changes_per_hour int4 := #{Settings.importer_changes_per_hour};
+      moderator_changes_per_hour int4 := #{Settings.moderator_changes_per_hour};
+      roles text[];
+      last_block timestamp without time zone;
+      first_change timestamp without time zone;
+      active_reports int4;
+      time_since_first_change double precision;
+      max_changes double precision;
+      recent_changes int4;
+    BEGIN
+      SELECT ARRAY_AGG(user_roles.role) INTO STRICT roles FROM user_roles WHERE user_roles.user_id = api_rate_limit.user_id;
+
+      IF 'moderator' = ANY(roles) THEN
+        max_changes := moderator_changes_per_hour;
+      ELSIF 'importer' = ANY(roles) THEN
+        max_changes := importer_changes_per_hour;
+      ELSE
+        SELECT user_blocks.created_at INTO last_block FROM user_blocks WHERE user_blocks.user_id = api_rate_limit.user_id ORDER BY user_blocks.created_at DESC LIMIT 1;
+
+        IF FOUND THEN
+          SELECT changesets.created_at INTO first_change FROM changesets WHERE changesets.user_id = api_rate_limit.user_id AND changesets.created_at > last_block ORDER BY changesets.created_at LIMIT 1;
+        ELSE
+          SELECT changesets.created_at INTO first_change FROM changesets WHERE changesets.user_id = api_rate_limit.user_id ORDER BY changesets.created_at LIMIT 1;
+        END IF;
+
+        IF NOT FOUND THEN
+          first_change := NOW();
```

Comment:
As mentioned in a previous comment, could you change all NOW() in this db function  to `now() at time zone 'utc'` to facilitate testing in non utc-0 environments.

Reply:
Is this because the DB columns are `timestamp` instead of `timestamptz` like `now()`?

Reply:
Probably yes. The database function wasn't working properly without "at time zone 'utc'".